### PR TITLE
CLIP Score Fusion model implementation

### PR DESCRIPTION
**Pull request by vinay-raman**
# What does this PR do ?

Implemented ClipScoreFusion model (UniIR) which leverages Clip models for multimodal retrieval tasks

**Collection**: [Note which collection this PR will affect]
multimodal

# Usage

```python
model = MegatronCLIPScoreFusionModel(cfg.model, trainer)
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?
@JRD971000

### CODE REVIEW

File: nemo/collections/multimodal/models/vision_language_foundation/clip/megatron_clip_scorefusion_models.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    def __init__(self, cfg: DictConfig, trainer: Trainer, pre_process=True, post_process=True):
+        super().__init__(cfg, trainer)
+
+    #     #TODO add support for huggingface models instead of 
+    #                           conversion to .nemo
+       
+        self.tokenizer = clip.tokenize
+        
+
+        # self.logit_scale =
```
Comment (JRD971000) [inline nemo/collections/multimodal/models/vision_language_foundation/clip/megatron_clip_scorefusion_models.py:61]: remove comment

File: nemo/collections/multimodal/models/vision_language_foundation/clip/megatron_clip_scorefusion_models.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
+import torch
```
Comment (JRD971000) [inline nemo/collections/multimodal/models/vision_language_foundation/clip/megatron_clip_scorefusion_models.py:1]: add licensing head to all the files

File: nemo/collections/multimodal/data/clip/mbeir_dataset.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+                    neg_cand.get("img_path", None),
+                )
+                for neg_cand in selected_neg_cand_list
+            ]
+            if len(neg_cand_list) > 0:
+                instance.update({"neg_cand_list": neg_cand_list})
+        return instance
+
+
+class MBEIRCandidatePoolDataset(MBEIRDatasetBase):
```
Comment (JRD971000) [inline nemo/collections/multimodal/data/clip/mbeir_dataset.py:277]: If these dataset are fully/partially copied from an external repo, we have to make sure about their license

File: examples/multimodal/vision_language_foundation/clip/convert_external_clip_to_nemo.py [hunk 4]
```diff
@@ -168,8 +168,8 @@ def mapping_hf_state_dict(hf_model):
         ".pre_layrnorm.bias": ".preprocess_layernorm.bias",
         ".post_layernorm.weight": ".transformer.final_layernorm.weight",
         ".post_layernorm.bias": ".transformer.final_layernorm.bias",
-        ".backbone.embeddings.position_embedding.weight": ".backbone.position_embeddings",
-        ".language_model.embeddings.position_embedding.weight": ".language_model.embedding.position_embeddings",
+        ".backbone.embeddings.position_embedding.weight": ".backbone.position_embeddings.weight",
```
Comment (JRD971000) [inline examples/multimodal/vision_language_foundation/clip/convert_external_clip_to_nemo.py:171]: @yaoyu-33 Do we have a test for this conversion in CICD?

File: examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  data_config:
+      image_size: 224, 224
+      hard_neg_num: 0
+      in_batch_neg_num: 0  # TODO: Move this to model config
+      shuffle_cand: True
+      
+      returns: null
+
+      # Relative to mbeir_data_dir
+      mbeir_data_dir: "/data/M-BEIR"
```
Comment (JRD971000) [inline examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml:237]: remove any directory path

File: examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+      enable_query_instruct: True
+      query_instruct_path: "instructions/query_instructions.tsv"
```
Comment (JRD971000) [inline examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml:239]: same

File: examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+      train_query_data_path: "/data/M-BEIR/query/union_train/mbeir_union_up_train.jsonl"
```
Comment (JRD971000) [inline examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml:241]: same

File: examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+  precision: ${trainer.precision}
+  # specify micro_batch_size, global_batch_size, and model parallelism
+  # gradient accumulation will be done automatically based on data_parallel_size
+  micro_batch_size: 105 # limited by GPU memory
+  global_batch_size: 105 # will use more micro batches to reach global batch size
+  tensor_model_parallel_size: 1 # intra-layer model parallelism
+  pipeline_model_parallel_size: 1 # inter-layer model parallelism
+  virtual_pipeline_model_parallel_size: null # interleaved pipeline
+
+  restore_from_path: /NeMo/clip_test/vitb32.nemo
```
Comment (JRD971000) [inline examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml:60]: remove directory path

File: nemo/collections/multimodal/data/clip/mbeir_dataset.py [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+        }
+
+        # Print dataset config
+        if print_config:
+            self.cand_pool_path = cand_pool_data_path
+            self._print_config()
+
+    def _print_config(self):
+        # Print dataset config
+        print(f"\n---Mbeir Candidate Pool Dataset Config---")
```
Comment (JRD971000) [inline nemo/collections/multimodal/data/clip/mbeir_dataset.py:303]: replace print with logging, all nemo models use that and we avoid using print (sorry for nit picking :) )

**Review by JRD971000 — COMMENTED** [review]
left a few comments

File: nemo/collections/multimodal/models/vision_language_foundation/clip/megatron_clip_scorefusion_models.py (hunk 1 above)
Comment (vinay-raman) [reply nemo/collections/multimodal/models/vision_language_foundation/clip/megatron_clip_scorefusion_models.py:61]: fixed thanks!

File: nemo/collections/multimodal/data/clip/mbeir_dataset.py (hunk 3 above)
Comment (vinay-raman) [reply nemo/collections/multimodal/data/clip/mbeir_dataset.py:277]: This is from [uniir](https://github.com/TIGER-AI-Lab/UniIR/tree/main/src/data), its MIT license. I can add the link in the comments in the file.

File: examples/multimodal/vision_language_foundation/clip/convert_external_clip_to_nemo.py (hunk 4 above)
Comment (vinay-raman) [reply examples/multimodal/vision_language_foundation/clip/convert_external_clip_to_nemo.py:171]: I have tested it myself, its not in CICD, I was told that someone was working on a fix for this and there was a separate MR for fixing bugs in conversion. I will find out and get back. If its not there I will add it to CICD

File: examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml (hunk 5 above)
Comment (vinay-raman) [reply examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml:237]: fixed, thanks!

File: examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml (hunk 7 above)
Comment (vinay-raman) [reply examples/multimodal/vision_language_foundation/clip/conf/megatron_clipscorefusion_config.yaml:241]: fixed, thanks!

**Pull request closed without merging**